vault/diagnose/tls_verification.go

```diff
@@ -28,64 +37,119 @@ func ListenerChecks(listeners []listenerutil.Listener) error {
 		}
 		_, ok := tlsutil.TLSLookup[l.TLSMinVersion]
 		if !ok {
-			return fmt.Errorf(minVersionError, l.TLSMinVersion)
+			err := fmt.Errorf("listener at address: %s has error %s: ", listenerID, fmt.Sprintf(minVersionError, l.TLSMinVersion))
+			listenerErrors = append(listenerErrors, err)
+			Error(ctx, err)
 		}
 		_, ok = tlsutil.TLSLookup[l.TLSMaxVersion]
 		if !ok {
-			return fmt.Errorf(maxVersionError, l.TLSMaxVersion)
+			err := fmt.Errorf("listener at address: %s has error %s: ", listenerID, fmt.Sprintf(maxVersionError, l.TLSMaxVersion))
+			listenerErrors = append(listenerErrors, err)
+			Error(ctx, err)
 		}
 
 		// Perform checks on the TLS Cryptographic Information.
-		if err := TLSFileChecks(l.TLSCertFile, l.TLSKeyFile); err != nil {
-			return err
+		warnings, err := TLSFileChecks(l.TLSCertFile, l.TLSKeyFile)
+		for _, warning := range warnings {
+			warning = listenerID + ": " + warning
+			listenerWarnings = append(listenerWarnings, warning)
+			Warn(ctx, warning)
 		}
+		if err != nil {
+			errMsg := listenerID + ": " + err.Error()
+			listenerErrors = append(listenerErrors, fmt.Errorf(errMsg))
+			Error(ctx, fmt.Errorf(errMsg))
+		}
+
+		// TODO: Use listenerutil.TLSConfig to warn on incorrect protocol specified
+		// Alternatively, use tlsutil.SetupTLSConfig.
 	}
-	return nil
+	return listenerWarnings, listenerErrors
+}
+
+// TLSFileChecks returns an error and warnings after checking TLS information
+func TLSFileChecks(certpath, keypath string) ([]string, error) {
+	// Parse TLS Certs from the certpath
+	leafCerts, interCerts, rootCerts, err := ParseTLSInformation(certpath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check for TLS Warnings
+	warnings, err := TLSFileWarningChecks(leafCerts, interCerts, rootCerts)
+	if err != nil {
+		return warnings, err
+	}
+
+	// Check for TLS Errors
+	if err = TLSErrorChecks(leafCerts, interCerts, rootCerts); err != nil {
+		return warnings, err
+	}
+
+	// Utilize the native TLS Loading mechanism to ensure we have missed no errors
+	_, err = tls.LoadX509KeyPair(certpath, keypath)
+	return warnings, err
 }
 
-// TLSFileChecks contains manual error checks against the TLS configuration
-func TLSFileChecks(certFilePath, keyFilePath string) error {
+// ParseTLSInformation parses certficate information and returns it from a cert path.
+func ParseTLSInformation(certFilePath string) ([]*x509.Certificate, []*x509.Certificate, []*x509.Certificate, error) {
+	leafCerts := []*x509.Certificate{}
+	interCerts := []*x509.Certificate{}
+	rootCerts := []*x509.Certificate{}
 	data, err := ioutil.ReadFile(certFilePath)
 	if err != nil {
-		return fmt.Errorf("failed to read tls_client_ca_file: %w", err)
+		return leafCerts, interCerts, rootCerts, fmt.Errorf("failed to read tls_client_ca_file: %w", err)
```

Comment:
I think certFilePath is for tls_cert_file, the error, however, states tls_client_ca_file.

Reply:
Oops, my mistake -- good catch!